src/Pull.cpp

```diff
@@ -0,0 +1,123 @@
+#include "Pull.h"
+#include <stdexcept>
+#include <g3log/g3log.hpp>
+/**
+ * Construct a blocking vampire (no timeout)
+ */
+Pull::Pull(const std::string& location, const bool shouldConnect) {
+   mProtocolHandler = std::move(std::unique_ptr<NanoProtocol>(
+                                   new NanoProtocol(location)));
+   mSocket = nn_socket(AF_SP, NN_PULL);
+   if (mSocket < 0) {
+      throw std::runtime_error("could not open transport socket");
+   }
+   int connectResponse = {0};
+   std::string connectionType("");
+   if (shouldConnect){
+      connectResponse = nn_connect(mSocket,
+                                   mProtocolHandler->GetLocation().c_str());
+      connectionType = "nn_connect";
+   } else{
+      connectResponse = nn_bind(mSocket,
+                                mProtocolHandler->GetLocation().c_str());
+      connectionType = "nn_bind";
+   }
+   if (connectResponse < 0) {
+      throw std::runtime_error("could not connect to endpoint: " +
```

Comment:
I think a throwing constructor is OK. We usually don't do it this way but it's OK. 

the more common pattern is to have a factory method that calls the internals, if the internal are OK they are passed to the private constructor and the object is return in a unique ptr. If the internals are not OK  a nullptr unique_ptr is returned

I prefer the factory method but don't see anything "wrong" with this way of handling it either as long as the client code using the Pull has exception handling mechanisms